pkg/shell/state.tsx

```diff
@@ -17,121 +17,191 @@
-    let self = null;
+interface ShellConfig {
+    language: string;
+    language_direction: string;
+    host_switcher_enabled: boolean;
+}
+
+interface ShellPageStatus {
+    type: string; // ???
+    message: string; // ???
+}
 
-    /* CONFIG
+interface ShellFrame {
+    name: string;
+    host: string;
+    path: string;
+    title: string;
+    url: string | null;
+    hash: string;
+    ready: boolean;
+    loaded: boolean;
+}
+
+interface ShellStateEvents {
+    update: () => void;
+    connect: () => void;
+}
+
+export class ShellState extends EventEmitter<ShellStateEvents> {
+    constructor() {
+        super();
+        this.config = this.#init_config();
+
+        this.machines = this.#init_machines();
+        this.loader = this.#init_loader();
+        this.router = this.#init_router();
+
+        this.#init_watch_dogs();
+        this.#init_page_status();
+
+        this.#on_ready();
+    }
+
+    /* READINESS STATE
      */
 
-    let language = document.cookie.replace(/(?:(?:^|.*;\s*)CockpitLang\s*=\s*([^;]*).*$)|^.*$/, "$1");
-    if (!language)
-        language = navigator.language.toLowerCase(); // Default to Accept-Language header
-
-    const config = {
-        language,
-        language_direction: cockpit.language_direction,
-        host_switcher_enabled: false,
-    };
-
-    /* Host switcher enabled? */
-    const meta_multihost = document.head.querySelector("meta[name='allow-multihost']");
-    if (meta_multihost instanceof HTMLMetaElement && meta_multihost.content == "yes")
-        config.host_switcher_enabled = true;
-
-    /* Should show warning before connecting? */
-    let config_ready = false;
-    cockpit.dbus(null, { bus: "internal" }).call("/config", "cockpit.Config", "GetString",
-                                                 ["Session", "WarnBeforeConnecting"], {})
-            .then(([result]) => {
-                if (result == "false" || result == "no") {
-                    window.sessionStorage.setItem("connection-warning-shown", "yes");
-                }
-            })
-            .catch(e => {
-                if (e.name != "cockpit.Config.KeyError")
-                    console.warn("Error reading WarnBeforeConnecting configuration:", e.message);
-            })
-            .finally(() => {
-                config_ready = true;
-                on_ready();
+    ready: boolean = false;
+    problem: string | null = null;
+    has_oops: boolean = false;
+
+    #on_ready() {
+        if (this.machines.ready && this.#config_ready) {
+            this.ready = true;
+            window.addEventListener("popstate", () => {
+                this.update();
+                this.ensure_frame_loaded();
+                this.ensure_connection();
             });
 
-    /* MACHINES DATABASE AND MANIFEST LOADER
+            this.update();
+            this.ensure_frame_loaded();
+            this.ensure_connection();
+        }
+    }
+
+    /* CONFIG
+     */
+
+    config: ShellConfig;
```

Comment:
This syntax looks strange to me. So this declares the type, but it doesn't assign any value.. I suppose it would start as `undefined`, but how does that not break the assertion that `config` is always a `ShellConfig`? As `ShellConfig` is just an interface, it could not even implicitly create an object and calls the constructor (and doing so would also be rather evil).

Does tsc make sure that this member gets initialized in the constructor? That would be plausible, clever, and robust.

:thinking:

Reply:
When commenting out the call in the ctor, tsc indeed fails:

> config: ShellConfig;     ■ Property 'config' has no initializer and is not definitely assigned in the constructor.

So indeed this is the case.